### PR TITLE
[Tests] Add a test to round-trip types through mangled names.

### PR DESCRIPTION
**Pull request by al45tair**
Added a new test to the test suite that round trips Swift types through mangled names and checks that we get the same type back that we started with.

rdar://37170485

### CODE REVIEW

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
@swift-ci Please smoke test

File: test/TypeRoundTrip/Inputs/RoundTrip/RoundTrip.cpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  Demangle::StackAllocatedDemangler<1024> dem;
+  auto nodeTree = _swift_buildDemanglingForMetadata(md, dem);
+  if (!nodeTree)
+    return "<unknown>";
+
+  std::string result = Demangle::nodeToString(nodeTree);
+  return result;
+}
+
+extern "C" void roundTripType(const Metadata *md) {
```
Comment (DougGregor) [inline test/TypeRoundTrip/Inputs/RoundTrip/RoundTrip.cpp:24]: Use the Swift calling convention here?

**Review by DougGregor — APPROVED** [review]

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
@swift-ci Please smoke test Windows platform

File: test/TypeRoundTrip/Inputs/RoundTrip/RoundTrip.cpp (hunk 1 above)
Comment (al45tair) [reply test/TypeRoundTrip/Inputs/RoundTrip/RoundTrip.cpp]: Done.

**Pull request merged**